feat: added Why TypeScript Doesn't Follow Strict Semantic Versioning article (#114)
LearningTypeScript/site: src/content/articles/why-typescript-doesnt-follow-strict-semantic-versioning.mdx

---
date: "July 20 2023"
description: "Why it's not practically feasible -or desirable- for a programming language like TypeScript to use strict semantic versioning for its releases."
meta: ECMAScript, compiler options target, TSConfig
---

import { CaptionedImage } from "@site/src/theme/CaptionedImage";

# Why TypeScript Doesn't Follow Strict Semantic Versioning

> :::danger

Reviewer: Do you remove this before publishing after folks have gotten back to you? Seems to undermine the purpose of your blog post :\ 

I think you can opt for a less critical intro/disclaimer. Something like:
I was inspired to write this blog after I read some hot takes on TS semantic versioning. This is not meant to put an end to the discussion but provide some clarity and offer my own perspective. If you have any feedback on how to improve this post, please let me know!

Author: I do remove this before publishing, yes 😄 but thanks!

Changes to these definitions can cause type errors in existing code, even if the changes are purely additive.
Many projects augment built-in type definitions to fill in gaps or make them more strict per the project's needs.
Modifications to augmented built-in types can introduce type errors for [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) that previously was permitted.

Example: [TypeScript 4.9 improved the accuracy of the built-in `Promise.resolve` types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#better-types-for-promiseresolve).
That improved type can break existing code relying on `Promise.resolve(...)` calls previously typed as `any` or `unknown` (&lt;4.9) instead of a `Promise<...>` (>=4.9).

## Strict Versioning Options

Given how "breaking" many common type checking changes are, it follows that if TypeScript followed a strict interpretation of semver, _every release would be a major_.
TypeScript's version strategy options therefore are:

- Allowing changes to the type system in non-major versions _(the strategy in use today)_
- Only releasing those changes in major versions

> Unfortunately, only releasing type system changes in major versions is impractical regardless of how often those major versions are released.

Reviewer: Is that to say it’s impractical because of how often major versions are released? (instead of "regardless of")

Author: Good clarity point - reordering...

However, doing so would negate much of the benefits of semver for consumers.
With a major version released every 3 months, consumers would have to update to that major version very frequently.
The distinction between major and minor versions would lose meaning.

That loss of distinction would be detrimental to consumers because TypeScript does occasionally need to release an API-level breaking change.
If every release were to be branded as major, then neither developers nor tooling would be able to easily identify the "true" major breakages.

### Infrequent Majors

On the other hand, TypeScript could release major versions infrequently - say, once every six months, or even every year.
This would minimize the frequency consumers would need to update major versions and experience breakages.

But restricting all "breaking" changes to an occasional new major version would mean bottling up all changes for months at a time.
New infrequent major versions would contain a significant amount of type system changes all at once.
Onboarding to new infrequent major versions would be much more difficult than today's minor versions.

> Even worse, because the new changes would have so little public user testing in stable versions versions before release, they'd likely contain ma ny more bugs.

Reviewer: lol, maybe this was a joke, but "ma ny" => "many"

Author: heh nope

1. Fix small issues with the types to work with the new TypeScript behavior
2. Refactor the types to reduce complexity and work with the new TypeScript behavior
3. Use `any` on types you can't figure out, with a `// TODO` comment linked to a tracking issue/ticket for cleaning it up later
4. Use `// @ts-expect-error` on remaining issues you can't figure out, with a `// TODO` comment linked to a tracking issue/ticket for cleaning it up later

Let's look at those tips in more details.

#### Straightforward, Clean Types

TypeScript's type system is fantastically powerful and can represent some wild and wacky type operations.
However, the more complex your types are, the more difficult they are to read, write, and update over time.
TypeScript types follow simplicity principles from other types of code:

- Prefer simple, easy-to-read code (types) whenever possible
- Prefer well-named types that explain what they're doing
- When code does become complex to the point of being difficult to read, consider using comments to explain the tricky details

Simpler types are less likely to be broken by new TypeScript versions tinkering with nuances of the type system.

> They're also easier to debug when something becomes broken.

Reviewer: This could be a chance to toot your horn, consider linking to a project (doesn't have to be yours either) that might exemplify these patterns.

Author: Ha, I think I'll pass - I have too much bashfulness and suspect others do too.

The `any` type indicates that a value that can be <em>any</em>thing and TypeScript should allow using it in almost any way.
Which is as dangerous as it sounds: `any` stops TypeScript from reporting potentially useful type errors and stops other [TypeScript developer utilities](https://code.visualstudio.com/docs/languages/typescript#_code-navigation) from working as well.

Still, if you're struggling to get a type to work, `any` can be an effective backup band-aid to lessen TypeScript's type checking strictness.
Replacing a type with `any` can save you from having to dive into complexities around that type.
Try to always add a `// TODO` comment explaining why you used the `any`, to help inform future efforts to remove the `any`.

:::tip
The [`@typescript-eslint/no-explicit-any`](https://typescript-eslint.io/rules/no-explicit-any) lint rule can enforce against explicitly writing `any` types in code.
:::

#### Falling Back to TypeScript Comment Directives

Even more dangerous than `any` are the TypeScript _comment directives_ that completely turn off the type checker for a line:

- `// @ts-ignore`: Silences any type checking errors for a line

> - `// @ts-expect-error`: Acts like `// @ts-ignore`, but if a type error isn't reported, causes a new type error for the unnecessary comment

Reviewer: I was tripped up by this wording. After re-reading and getting more context from the following paragraphs, consider:
"Acts like ... but if the corresponding line doesn't actually throw an error, TS will show an error that this comment is necessary"

These are absolute last ditch strategies for when all else has failed.
Silencing TypeScript type errors altogether brings the same downside as an `any`, but even more so - across an entire line of code.

If you absolutely must use a comment directive to silence TypeScript, prefer using `// @ts-expect-error` with a `// TODO` comment explaining why you needed it.
If a future change to your codebase removes the need for a comment directive, the `// @ts-expect-error` will direct TypeScript to let you know to remove the comment.

:::tip

> The [`@typescript-eslint/ban-ts-comment`](https://typescript-eslint.io/rules/ban-ts-comment) lint rule can enforce against comment directives, or enforce that they be accompanied by an explanatory comment

Reviewer: missing period

# Why TypeScript Doesn't Follow Strict Semantic Versioning

Most projects in the JavaScript/TypeScript ecosystem release new versions with numbers respecting to _**[semantic versioning](https://semver.org)**_, or _semver_ for short.
TypeScript is notable for not following a traditional interpretation of semver for its releases.
This article will dig into:

1. What semver is and why it's useful for many packages
2. Why following a strict interpretation of semver would be impractical for TypeScript
3. How TypeScript's releases are versioned to an interpretation of semver that makes sense for it

While TypeScript's diverging from a common community specification can be irksome for developers, there are real reasons why it chose to do so.

<!-- truncate -->

## Context: Semantic Versioning

Semver is a specification that describes how to increase a package's version numbers upon each new release.
Specifically, it dictates that versions adhere to a _Major.Minor.Patch_ format:

- _Major_: increases when you make incompatible API changes
- _Minor_: increases when you add functionality in a backward compatible manner
- _Patch_: increases when you make backward compatible bug fixes

> For example, given a package at version `1.2.3`, it might increase its version to:

Reviewer 2: [nitpick] I'm having trouble understanding this sentence. Maybe you can remove "For example"? 
```suggestion
For example, given a package at version `1.2.3`, it might increase its version to:
```

Author: That does, thanks. My instinct in cases where the sentence flow is unclear is to try the split the sentence in two. I have a tendency to write long sentences.

Author: ```suggestion
For example, suppose a package is at version `1.2.3`.
It might increase its version to:
```

## TypeScript Versioning Today

Given all the advantages of semantic versioning, it can be surprising to learn that TypeScript intentionally does not follow _strict_ semantic versioning.
TypeScript's version numbers to date have instead incremented according to the following pattern:

- A new _minor_ version of TypeScript (e.g. 4.9.0) is released every three months or so with new features
  - _Patch_ updates to that version (e.g. 4.9.1) which fix for small bugs are released as needed
- A new _major_ version of TypeScript (e.g. 5.0.0) is released if the minor version would have exceeded 9 (e.g. 4.10.0)

In other words, TypeScript uses _major_ version numbers more for marketing and public visibility than for semantic versioning.
This was an intentional engineering decision: releasing a new version of TypeScript almost always requires compiler changes that "break" some existing TypeScript programs.

### "Breaking" Type Checking Changes

Many users consider TypeScript's type checking as a part of its API.

> When a minor or patch version introduces changes to type checkin, some might consider that a "breaking change".

Reviewer 2: [Proofreading/question] Is this how checking is supposed to be spelled? 🤔 
```suggestion
When a minor or patch version introduces changes to type checking, some might consider that a "breaking change".
```

Author: aha, yeah - fixed locally

New versions of TypeScript are often made more capable in finding more classses of type errors.
But what some would consider bugfixes or new features in type checking, others would consider introduced breakages.

Updating to a new version of TypeScript and seeing new red editor squigglies or build failures could be considered "breaking" - but is a natural consequence of TypeScript improving over time.

Example: [TypeScript 4.8 stopped allowing unconstrained generic type parameters to be assignable to `{}`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-8.html#unconstrained-generics-no-longer-assignable-to-).
This is technically correct, as `{}` is not supposed to be assignable to `null` or `undefined`.
But existing code might have been written with the assumption that the type parameters would always be provided with an object type.
That code would newly receive type errors in TypeScript >=4.8.

#### Removed Type Errors

Similarly, TypeScript sometimes stops reporting type errors in situations that used to be considered incorrect.
This is sometimes done to make TypeScript more permissive and compatible with existing JavaScript code.

> But for consumers who relied on TypeScript's stricter type checking to catch issues, these changes could also be considered "breaking".

Reviewer 2: [Proofreading] I got the feeling that this statement describes consumers only using TS stricter type checking for this purpose, so maybe you can "solely" to make that point clearer. 
```suggestion
But for consumers who solely relied on TypeScript's stricter type checking to catch issues, these changes could also be considered "breaking".
```

Author: This would be inaccurate. You'd generally rely on a lot of things from TypeScript to catch issues, including strict and less-strict parts.

> Updating to a new version of TypeScript and no longer being prevented from shipping code patterns the project wanted to avoid could also be considered "breaking" - even if TypeScript considers the change an improvement.

Reviewer 2: [nitpick] What is this paragraph describing here? It could use some clarity. 
```suggestion
Updating to a new version of TypeScript and no longer being prevented from shipping code patterns the project wanted to avoid could also be considered "breaking" - even if TypeScript considers the change an improvement.
```

Additionally, there are a few strategies you can take to minimize that pain.

### Pinned TypeScript Versions

Most software projects in the JavaScript ecosystem today use a "lockfile" such as a [`package-lock.json` (npm)](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json), [`pnpm-lock.yaml` (pnpm)](https://pnpm.io/cli/install), and [`yarn.lock` (Yarn)](https://classic.yarnpkg.com/lang/en/docs/yarn-lock) to keep consistent versions of packages.
A lockfile will ensure that installing dependencies in your project will always use the same versions.
Using a lockfile is generally a good practice - especially for tools such as TypeScript that can substantially change behavior across versions.

Note that when using a lockfile, you don't have to specify an exact _major.minor.patch_ version for `"typescript"` in your `package.json`.
Your lockfile will specify an exact version of every dependency for you - generally defaulting to the latest stable version of any dependencies in your `package.json`, including `"devDependencies"`.

### Intentional TypeScript Version Upgrades

> When you do want to update your TypeScript version, consider the following strategy:

Reviewer 2: [Nitpick] This seems more like a set of steps instead of strategies.
```suggestion
When you do want to update your TypeScript version, consider the following steps:
```

Author: I like "strategy" :) it implies that this is a rough approach, not an explicit list of steps to take. Even if they are just steps I'd rather people see them as a strategy.
